From 979d92657b0ea27f00b664ccd40a4afa2f0fecc9 Mon Sep 17 00:00:00 2001
From: WeiguangTWK <weiguangtwk@outlook.com>
Date: Thu, 9 Jul 2026 14:06:57 +0800
Subject: [PATCH] Neko LineageOS: [Anti Detection] Tighten NTFS fsck SEPolicy

Some of detector (e.g. Duckdetector) detects custom roms through NTFS related over-loose SELinux rules

Change-Id: I3df7f9584913923ee37a8ee344bcecd1048082b5
---
 common/private/fsck_ntfs_untrusted.te | 53 +++++++++++++++++++++++++++
 common/private/fsck_untrusted.te      |  2 -
 common/private/vold.te                |  1 +
 3 files changed, 54 insertions(+), 2 deletions(-)
 create mode 100644 common/private/fsck_ntfs_untrusted.te
 delete mode 100644 common/private/fsck_untrusted.te

diff --git a/common/private/fsck_ntfs_untrusted.te b/common/private/fsck_ntfs_untrusted.te
new file mode 100644
index 0000000..54ab59d
--- /dev/null
+++ b/common/private/fsck_ntfs_untrusted.te
@@ -0,0 +1,53 @@
+type fsck_ntfs_untrusted, domain, coredomain;
+type fsck_ntfs_untrusted_exec, system_file_type, exec_type, file_type;
+
+# Allow stdin/out back to vold.
+allow fsck_ntfs_untrusted vold:fd use;
+allow fsck_ntfs_untrusted vold:fifo_file { read write getattr };
+
+# Inherit and use pty created by android_fork_execvp_ext().
+allow fsck_ntfs_untrusted devpts:chr_file { read write ioctl getattr };
+
+# Run fsck on vold block devices and stat all block devices to detect mounts.
+allow fsck_ntfs_untrusted block_device:dir search;
+allow fsck_ntfs_untrusted vold_device:blk_file rw_file_perms;
+allow fsck_ntfs_untrusted dev_type:blk_file getattr;
+allow fsck_ntfs_untrusted proc_mounts:file r_file_perms;
+
+# NTFS userspace fsck probes block device properties that can require CAP_SYS_ADMIN.
+allow fsck_ntfs_untrusted self:capability sys_admin;
+
+###
+### neverallow rules
+###
+
+# Untrusted NTFS fsck should never be run on block devices holding sensitive data.
+neverallow fsck_ntfs_untrusted {
+  boot_block_device
+  frp_block_device
+  metadata_block_device
+  recovery_block_device
+  root_block_device
+  swap_block_device
+  system_block_device
+  userdata_block_device
+  cache_block_device
+  dm_device
+}:blk_file no_rw_file_perms;
+
+# Only allow entry from vold via fsck.ntfs.
+neverallow { domain -vold } fsck_ntfs_untrusted:process transition;
+neverallow * fsck_ntfs_untrusted:process dyntransition;
+neverallow fsck_ntfs_untrusted { file_type fs_type -fsck_ntfs_untrusted_exec }:file entrypoint;
+
+# Similar to fsck_untrusted, the NTFS-specific helper never needs uid/gid-changing caps.
+neverallow fsck_ntfs_untrusted self:global_capability_class_set { setgid setuid };
+
+###
+### dontaudit rules
+###
+
+dontaudit fsck_ntfs_untrusted sysfs:file rw_file_perms;
+dontaudit fsck_ntfs_untrusted sysfs_dm:file rw_file_perms;
+dontaudit fsck_ntfs_untrusted sysfs_dm:dir rw_dir_perms;
+dontaudit fsck_ntfs_untrusted tmpfs:lnk_file read;
diff --git a/common/private/fsck_untrusted.te b/common/private/fsck_untrusted.te
deleted file mode 100644
index 5d12f76..0000000
--- a/common/private/fsck_untrusted.te
+++ /dev/null
@@ -1,2 +0,0 @@
-# External storage
-allow fsck_untrusted self:capability sys_admin;
diff --git a/common/private/vold.te b/common/private/vold.te
index 915190b..aa36188 100644
--- a/common/private/vold.te
+++ b/common/private/vold.te
@@ -2,6 +2,7 @@
 allow vold self:capability { setgid setuid };
 
 # External storage
+domain_trans(vold, fsck_ntfs_untrusted_exec, fsck_ntfs_untrusted);
 allow vold mkfs_exec:file rx_file_perms;
 allow vold mnt_media_rw_stub_file:dir r_dir_perms;
 allow vold storage_stub_file:dir rw_dir_perms;
-- 
2.43.0

